ford/sourceform.py

```diff
@@ -1055,39 +1097,92 @@ def filter_public(collection: dict) -> dict:
                 typelist[dtype] = set([])
         typeorder = toposort.toposort_flatten(typelist)
 
+        # Correlate types
+        for dtype in typeorder:
+            if dtype in self.types:
+                dtype.correlate(project)
+
         # Match up called procedures
         if hasattr(self, "calls"):
             tmplst = []
             for call in self.calls:
-                call_low = call.lower()
+                call_low = call["name"].lower()
                 argname = False
                 for a in getattr(self, "args", []):
                     # Consider allowing procedures passed as arguments to be included in callgraphs
                     argname |= call_low == a.name.lower()
                 if hasattr(self, "retvar"):
                     argname |= call_low == self.retvar.name.lower()
+                
+                # travers the call chain of the call to descover the context the call is made on
+                context = self
```

Comment:
Another bit that I think would be better pulled out into its own function. Then it will be easier to test it separately, as well as help keep this current function easier to understand